### ISSUE TITLE
Glue (Sled) pinning disk

### ISSUE THREAD

**Issue by KyGost** [gluesql/gluesql#514]
In the process of making & testing #513 I noticed that Glue pins the disk at full usage.
It seems to halt too, perhaps because of disk waits.

This occurs when inserting and doing operations on >200K rows (using SledStorage).

nb. MultiSQL is fine with this so it isn't a Sled issue.

**Comment by ever0de** [comment]
First of all, I decided that it was close to a bug, so I added a bug label.
Please let me know if there are "improvements" that you need an option for.(If you need to label improvements)

**Comment by KyGost** [comment]
I'd consider this a bug too.
Not sure what the cause is.

**Comment by earney** [comment]
Anyone recommend some rust profiling tools?  That may help in figuring out where the bottleneck is.

**Comment by ever0de** [comment]
> Anyone recommend some rust profiling tools? That may help in figuring out where the bottleneck is.

I have no experience with profiling, but...

- <https://nnethercote.github.io/perf-book/profiling.html>
- <https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb>

I think I usually use something like this:

**Comment by earney** [comment]
Here's a profiler that might be of interest:
https://github.com/EmbarkStudios/puffin if anyone is interested.

**Comment by earney** [comment]
![image](https://user-images.githubusercontent.com/1158898/164115907-6feb4c3b-f242-4fcd-b5dc-97a717e08369.png)

**Comment by KyGost** [comment]
It seems to me (without good backing) like these closures might be a pretty notable issue. Perhaps it could be done a better way.
https://github.com/gluesql/gluesql/blob/1cdcef9759f9d8cfeb208dc5b22d51ad6f63a1f8/core/src/executor/evaluate/mod.rs#L34-L39

**Comment by KyGost** [comment]
Hmm... Does the perf test look similar with something like:
```sql
SELECT 1
FROM A
INNER JOIN B
  ON B.fk = A.pk
GROUP BY A.pk
```

**Comment by earney** [comment]
Here's the perf report after I changed to the select 1 query above:
![image](https://user-images.githubusercontent.com/1158898/164773202-674a92ca-a5ba-4a75-a0a9-3c6452cff65a.png)

**Comment by KyGost** [comment]
Yeah so looks to me as though the closures are a problem and occur for all sorts of evals, which is a problem for grouping, and is used in aggregation, which is a big problem.

This could probably easily be fixed by just changing these high-use closures into functions (or, if desperate, macros).

**Comment by KyGost** [comment]
@panarch 

Probably an area with some really easy fixes that would massively boost perf.
